### DISCUSSION TITLE
Question about the start of NTRIP connection between the NTRIP server and the base station.

### DISCUSSION THREAD

**Discussion by Anibal-Quiroz-P5** [semuconsulting/pygnssutils#55] [Q&A] [answered]
As a way of learning about an NTRIP system, I want to implement a "super basic" NTRIP system consisting of a basic NTRIP server, a client (rover), and a single base station. The three components, NTRIP server, rover, and base station, would be implemented solely by code (neither the rover nor the base station would be physical, only code). The three codes would be implemented on my computer in my Visual Studio Code.
The functions of the NTRIP server would be to always listen for the rover to request to initiate an NTRIP connection. In that case, the rover (via NTRIP connection) would ask the NTRIP server to connect it to the base station to receive RTCM data (which will be hard-coded, meaning it will be a single fixed RTCM message implemented in the base station code). Another function of the NTRIP server would be to "establish an NTRIP connection between it and the base station." The last function of the NTRIP server would be to send the RTCM data received from the base station to the rover through the NTRIP connection that the rover had initiated.
The function of the pseudo base station would be to always listen, meaning waiting for a "request for an NTRIP connection from the NTRIP server," and the base station would have to "accept that NTRIP connection" and transmit the hard-coded RTCM data it has in its code.
I have already implemented the code for the "client" rover with the GNSSNTRIPClient class from the pygnssutils library, and it is working, meaning it connects via NTRIP to the NTRIP server.
My question is: with any class, function, or module of the pygnssutils library, " can I implement the code that initiates the NTRIP connection originating from my NTRIP server and destined for the base station (implemented by software)?"

**Comment by semuadmin** [comment]
Hi @Anibal-Quiroz-P5 

I'm not sure if I fully understand your intention, but synthesising viable RTCM messages from scratch in software (_as opposed to using an external RTK source or RTK-enabled GNSS receiver_) is not a trivial undertaking - it would require a consummate understanding of the physics and protocols underpinning RTK technology. If you're learning about NTRIP, I would recommend starting with one of the many public domain NTRIP services such as [rtk2go](http://rtk2go.com/) as your "NTRIP Server", and/or a suitable RTK-enabled GNSS recever such as the u-blox ZED-F9P as your "base station" and/or "rover" (_**NB:** I receive absolutely no commission or advertising revenue from u-blox or their resellers_).

In a base station / moving rover configuration, the base station effectively acts as the NTRIP Server to the rover. The base station could itself receive corrections from another third-party NTRIP server, but typically you'd want to calibrate your base station antenna (_using accurate surveying techniques_) to operate independently.

The pygnssutils provides two utilities which might help here:

1. The [gnssserver](https://github.com/semuconsulting/pygnssutils?tab=readme-ov-file#cli-usage---ntrip-mode) utility, running in NTRIP Caster mode, as your "base station" - **BUT NOTE THAT** this in intended to work in conjunction with an external UBX RTK receiver such as the ZED-F9P, which acts as the RTCM3 data source. The utility does NOT generate RTCM3 data from scratch.
1. The [gnssntripclient](https://github.com/semuconsulting/pygnssutils?tab=readme-ov-file#gnssntripclient) utility as your 'rover', which you're already using.

This [quick reference wiki](https://www.semuconsulting.com/gnsswiki/#DGPS) might be helpful for background information.

Hope this helps.

**Reply by Anibal-Quiroz-P5** [reply]
Semuadmin, first of all, thank you very much for answering so quickly, with your extensive and accurate response you clarified a lot of doubts that I had regarding this topic, so thank you again. Secondly, my intention to use a software-implemented base station with a hard-coded RTCM message was solely to see how the NTRIP server initiates an NTRIP connection with a base station and not caring if the data delivered by the base station was real or not, I simply wanted the NTRIP server to be the originator of the NTRIP connection to the base. That was the only objective of making a base station with a fixed rtcm value.
Now after understanding your clarifications, I have also been reading the articles by Lefebure and both in the link: "http://lefebure.com/articles/ntrip-caster-setup/" titled "NTRIP Caster Setup" as well as the link "http://lefebure.com/articles/ntrip-names/" titled "NTRIP Component Names" the author refers to 2 different components, one the NTRIP SERVER and the other the NTRIP CASTER and says that the NTRIP SERVER is the that receives correction data from "a single base station" and that an NTRIP CASTER, on the other hand, acts as a data concentrator or distributor within an NTRIP network and receives GNSS correction data from multiple sources, such as NTRIP servers, and transmits them to clients who request them.
Taking into account that I can get 4 "UBLOX EVK-F9P-01" modules (which can be configured as a base or as a rover and that transmit RTCM data through its serial port and that are also built with a ZED-F9P module) with them and with my computer I would like to build an NTRIP system with 3 base stations (several kilometers apart) and a rover.
Taking into account the last thing I told you and also Lefebure's definitions regarding NTRIP SERVER and NTRIP CASTER, I ask you if to implement the gnssserver utility, (the one that runs in NTRIP Caster mode), for my 3 base stations (3 modules "UBLOX EVK-F9P-01"), you would have to put a computer (for example raspberry pi) associated with each of those base stations, right? Because I understand that the pygnssutils gnssserver utility connects to the base through the serial port, that is, the computer on which the gnssserver utility is running has to be physically close to the base station and, that is, that computer would be able to access RTCM data from the serial port coming from the base, and that computer would have to deliver those RTCM messages through an NTRIP connection also generated by the gnssserver utility, right?
And if all of the above were correct, you would have 3 computers (each corresponding to a base station and running the gnssserver utility on each of them). And my last question is if the pygnssutils library has any utility that implements the CASTER NTRIP, that is, it implements the server in charge of receiving the correction data that each of the 3 base stations will transmit so that when the rover wants to connect to one of them 3, be this server, the CASTER NTRIP, which is responsible for initiating the NTRIP connection with the chosen base station and thus it can (through gnssserver) send the rtcm data to the CASTER NTRIP and it in turn sends it. send to the rover that requested it.
I know there were a lot of clarifications on my part, but I made them because I'm a little lost on this topic. Thanks for your help !!

**Comment by semuadmin** [comment, marked as answer]
Firstly, please appreciate that the NTRIP and RTCM protocols are _proprietary_ protocols, licensed and published by the [Radio Technical Commission for Maritime Services](https://rtcm.myshopify.com/pages/about-us). Some implementation details are effectively in the public domain (e.g. via libraries like RTKLIB) but the definitive documentation is only available via membership subscription or one-off payment of around $400. The pertinent documentation can be found here:

[RTCM 10410-3 RTCM DGNSS Standards](https://rtcm.myshopify.com/collections/differential-global-navigation-satellite-dgnss-standards/products/differential-gnss-package-both-of-the-current-standards-10402-3-and-10403-2)

[RTCM 10410-1 Standard for NTRIP 2.0](https://rtcm.myshopify.com/collections/differential-global-navigation-satellite-dgnss-standards/products/rtcm-10410-1-standard-for-networked-transport-of-rtcm-via-internet-protocol-ntrip-version-2-0-with-amendment-1-june-28-2011)

[NTRIP Client Best Practices](https://rtcm.myshopify.com/collections/differential-global-navigation-satellite-dgnss-standards/products/rtcm-paper-2023-sc104-1344-ntrip-client-devices-best-practices)

**Note, however, that NTRIP is not the _only_ way to pass RTCM data between a base station and a rover** - you could simply use Long Range Radio (LoRA) to transmit the raw RTCM data over what is effectively an extended serial connection. This is probably the simpler and more common configuration, but it requires dedicated LoRA hardware. Note there are packages available from providers such as [Ardusimple](https://www.ardusimple.com/rtk-starter-kits/) which include all the relevant components as a collection of plug-in modules.

But I'm assuming you're only interested in an NTRIP RTK implementation for the purposes of this discussion...

NTRIP terminology can be a little confusing. The term "NTRIP Server" refers to the element which provides the RTCM data feed, while the term "NTRIP Caster" essentially refers to the web server which collates the various RTCM data feeds and broadcasts them over the public internet. NTRIP uses the standard HTTP protocol (i.e. TCP sockets) for data transmission, so the connection itself is fairly simple to implement using standard Python Socket or HTTP libraries. The `gnssserver` utility acts as _both_ NTRIP Server _and_ NTRIP Caster.

In answer to your questions:

1. To use the `gnssserver` library as an NTRIP 'base station', then  - yes - you'd need a computer capable of running Python 3 at each base station location, physically connected to a suitably-configured RTK-compatible GNSS receiver via its serial port (USB or UART), and also connected to the public internet using a static IP address (e.g. via WiFi or Ethernet). A Rasperry Pi running headless Raspberry Pi OS would be very suitable. The procedure is detailed in the [README](https://github.com/semuconsulting/pygnssutils?tab=readme-ov-file#cli-usage---ntrip-mode).
2. If you want your rover to act as an NTRIP client, the rover will need to be capable of connecting to the basestation over the public internet. This typically means it will need some kind of WiFi or 4G/5G internet connection, an NTRIP client such as `gnssntripclient`, and a direct physical connection to the rover GNSS receiver's serial port. 
3. To connect to the base station, the rover will need the public IP address of the basestation, the name of the mountpoint (which in the case of `gnssserver` is always 'pygnssutils'), and the authentication userid and password (which default to 'anon' and 'password'). FYI the connection quality and distance to the basestation can be reported via a UBX NAV-RELPOSNED message.

Hope this helps.

**Comment by Anibal-Quiroz-P5** [comment]
Thank you again for answering and being so clear in your answers, everything you told me clarified the doubts I had, and now I am no longer lost on this topic.
